core/primitives/src/reed_solomon.rs

```diff
@@ -56,3 +62,180 @@ pub fn reed_solomon_decode<T: BorshDeserialize>(
+    fn deserialize_single_part(data: &[u8]) -> std::io::Result<Self> {
+        Self::try_from_slice(data)
+    }
+}
+
+impl ReedSolomonEncoder {
+    pub fn new(total_parts: usize, ratio_data_parts: f64) -> ReedSolomonEncoder {
+        let rs = if total_parts > 1 {
+            let data_parts = reed_solomon_num_data_parts(total_parts, ratio_data_parts);
+            Some(ReedSolomon::new(data_parts, total_parts - data_parts).unwrap())
+        } else {
+            None
+        };
+        Self { rs }
+    }
+
+    pub fn total_parts(&self) -> usize {
+        match self.rs {
+            Some(ref rs) => rs.total_shard_count(),
+            None => 1,
+        }
+    }
+
+    pub fn data_parts(&self) -> usize {
+        match self.rs {
+            Some(ref rs) => rs.data_shard_count(),
+            None => 1,
+        }
+    }
+
+    pub fn encode<T: ReedSolomonEncoderSerialize>(
+        &self,
+        data: &T,
+    ) -> (Vec<ReedSolomonPart>, usize) {
+        match self.rs {
+            Some(ref rs) => reed_solomon_encode(rs, data),
+            None => {
+                let bytes = T::serialize_single_part(&data).unwrap();
+                let size = bytes.len();
+                (vec![Some(bytes.into_boxed_slice())], size)
+            }
+        }
+    }
+
+    pub fn decode<T: ReedSolomonEncoderDeserialize>(
+        &self,
+        parts: &mut [ReedSolomonPart],
+        encoded_length: usize,
+    ) -> Result<T, std::io::Error> {
+        match self.rs {
+            Some(ref rs) => reed_solomon_decode(rs, parts, encoded_length),
+            None => {
+                if parts.len() != 1 {
+                    return Err(std::io::Error::other(format!(
+                        "Expected single part, received {}",
+                        parts.len()
+                    )));
+                }
+                let Some(part) = &parts[0] else {
+                    return Err(std::io::Error::other("Received part is not expected to be None"));
+                };
+                T::deserialize_single_part(part.as_ref())
+            }
+        }
+    }
+}
+
+pub struct ReedSolomonEncoderCache {
+    ratio_data_parts: f64,
+    instances: HashMap<usize, Arc<ReedSolomonEncoder>>,
+}
+
+impl ReedSolomonEncoderCache {
+    pub fn new(ratio_data_parts: f64) -> Self {
+        Self { ratio_data_parts, instances: HashMap::new() }
+    }
+
+    pub fn entry(&mut self, total_parts: usize) -> Arc<ReedSolomonEncoder> {
+        self.instances
+            .entry(total_parts)
+            .or_insert_with(|| {
+                Arc::new(ReedSolomonEncoder::new(total_parts, self.ratio_data_parts))
+            })
+            .clone()
+    }
+}
+
+pub struct ReedSolomonPartsTracker<T> {
+    parts: Vec<ReedSolomonPart>,
+    encoded_length: usize,
+    data_parts_present: usize,
+    encoder: Arc<ReedSolomonEncoder>,
+    total_parts_size: usize,
+    phantom: std::marker::PhantomData<T>,
```

Comment:
what is the phantom for?

Reply:
this is needed to make rust happy with the generic type that is only used in `impl` and not the definition of the struct